scapy/contrib/loraphy2wan.py

```diff
@@ -0,0 +1,681 @@
+        return "", p
+
+
+class FCtrl_UpLink(Packet):
+    name = "FCtrl_UpLink"
+    fields_desc = [BitField("ADR", 0, 1),
+                   BitField("ADRACKReq", 0, 1),
+                   BitField("ACK", 0, 1),
+                   BitField("ClassB", 0, 1),
+                   BitFieldLenField("FOptsLen", 0, 4)]
+
+    # pylint: disable=R0201
+    def extract_padding(self, p):
+        return "", p
+
+
+class DevAddrElem(Packet):
+    name = "DevAddrElem"
+    fields_desc = [XByteField("NwkID", 0x0),
+                   LEX3BytesField("NwkAddr", b"\x00" * 3)]
+
+
+CIDs_up = {0x01: "ResetInd",
+           0x02: "LinkCheckReq",
+           0x03: "LinkADRReq",
+           0x04: "DutyCycleReq",
+           0x05: "RXParamSetupReq",
+           0x06: "DevStatusReq",
+           0x07: "NewChannelReq",
+           0x08: "RXTimingSetupReq",
+           0x09: "TxParamSetupReq",  # LoRa 1.1 specs
+           0x0A: "DlChannelReq",
+           0x0B: "RekeyInd",
+           0x0C: "ADRParamSetupReq",
+           0x0D: "DeviceTimeReq",
+           0x0E: "ForceRejoinReq",
+           0x0F: "RejoinParamSetupReq"}  # end of LoRa 1.1 specs
+
+
+CIDs_down = {0x01: "ResetConf",
+             0x02: "LinkCheckAns",
+             0x03: "LinkADRAns",
+             0x04: "DutyCycleAns",
+             0x05: "RXParamSetupAns",
+             0x06: "DevStatusAns",
+             0x07: "NewChannelAns",
+             0x08: "RXTimingSetupAns",
+             0x09: "TxParamSetupAns",  # LoRa 1.1 specs here
+             0x0A: "DlChannelAns",
+             0x0B: "RekeyConf",
+             0x0C: "ADRParamSetupAns",
+             0x0D: "DeviceTimeAns",
+             0x0F: "RejoinParamSetupAns"}  # end of LoRa 1.1 specs
+
+
+class ResetInd(Packet):
+    name = "ResetInd"
+    fields_desc = [ByteField("Dev_version", 0)]
+
+
+class ResetConf(Packet):
+    name = "ResetConf"
+    fields_desc = [ByteField("Serv_version", 0)]
+
+
+class LinkCheckReq(Packet):
+    name = "LinkCheckReq"
+    fields_desc = []
+
+
+class LinkCheckAns(Packet):
+    name = "LinkCheckAns"
+    fields_desc = [ByteField("Margin", 0),
+                   ByteField("GwCnt", 0)]
+
+
+class DataRate_TXPower(Packet):
+    name = "DataRate_TXPower"
+    fields_desc = [XBitField("DataRate", 0, 4),
+                   XBitField("TXPower", 0, 4)]
+
+
+class Redundancy(Packet):
+    name = "Redundancy"
+    fields_desc = [XBitField("RFU", 0, 1),
+                   XBitField("ChMaskCntl", 0, 3),
+                   XBitField("NbTrans", 0, 4)]
+
+
+class LinkADRReq(Packet):
+    name = "LinkADRReq"
+    fields_desc = [DataRate_TXPower,
+                   XShortField("ChMask", 0),
+                   Redundancy]
+
+
+class LinkADRAns_Status(Packet):
+    name = "LinkADRAns_Status"
+    fields_desc = [BitField("RFU", 0, 5),
+                   BitField("PowerACK", 0, 1),
+                   BitField("ChannelMaskACK", 0, 1)]
+
+
+class LinkADRAns(Packet):
+    name = "LinkADRAns"
+    fields_desc = [LinkADRAns_Status]
```

Comment:
This is not correct, you need to write this instead to be able to dissect and build the packet:
`fields_desc = [PacketField("status", LinkADRAns_Status(), LinkADRAns_Status)]`

Reply:
Corrected